File: ben/haskell/Q11-20-lists-cont/p12.hs
{-
Decode a run-length encoded list.
Given a run-length code list generated as specified in problem 11. Construct its uncompressed version.
-}

data Encoded = Single Char
             | Multiple Int Char
    deriving Show

decodeModified :: [Encoded] -> String
decodeModified [] = []
decodeModified (Single c:xs) = c : decodeModified xs
decodeModified (Multiple nb c:xs) = (replicate nb c) ++ decodeModified xs

File: ben/prolog/P01-P28-lists/p12.pl
% Decode a run-length encoded list.

decode([], []).
decode([X|Xs], [X|R]) :-
    \+ is_list(X),
    decode(Xs, R).
decode([[0,_]|Xs], R) :-
    decode(Xs, R).
decode([[N0,X]|Xs], [X|R]) :-
    N0 > 0,
    N1 is N0 - 1, 
    decode([[N1,X]|Xs], R).
